File: adventofcode2022/06.rs
fn star1() {
    let txt = std::fs::read_to_string("06.txt").unwrap();
    let mut prev = [0 as char; 4];
    let mut previ = 0;
    for (i, c) in txt.chars().enumerate() {
        prev[previ] = c;
        previ = (previ+1) % prev.len();
        //println!("{i} = {c}");
        if i > 3 {
            let mut has_same = false;
            for j in 0..prev.len() {
                for k in j+1..prev.len() {
                    if prev[j] == prev[k] {
                        has_same = true;
                    }
                }
            }
            if !has_same {
                println!("{}", i+1);
                break;
            }
        }
    }
    // 1274 too low
}

fn star2() {
    let txt = std::fs::read_to_string("06.txt").unwrap();
    let mut prev = [0 as char; 14];
    let mut previ = 0;
    for (i, c) in txt.chars().enumerate() {
        prev[previ] = c;
        previ = (previ+1) % prev.len();
        //println!("{i} = {c}");
        if i > 3 {
            let mut has_same = false;
            for j in 0..prev.len() {
                for k in j+1..prev.len() {
                    if prev[j] == prev[k] {
                        has_same = true;
                    }
                }
            }
            if !has_same {
                println!("{}", i+1);
                break;
            }
        }
    }
}

fn main() {
    star1();
    star2();
}

File: adventofcode2024/06.go
package main
import "fmt"
import "strings"
import "os"

func star1() {
	inBytes,_ := os.ReadFile("06.txt")
	in := string(inBytes)

	area := make([][]byte, 0)
	
	for _,line := range strings.Split(in, "\n") {
		if strings.TrimSpace(line) != "" {
			area = append(area, []byte(strings.TrimSpace(line)))
		}
	}

	ylen := len(area)
	xlen := len(area[0])

	

	gx, gy := -1,-1
	for y := 0; y < ylen; y++ {
		for x := 0; x < xlen; x++ {
			//fmt.Printf("%c", rune(area[y][x]))
			if area[y][x] == '^' { gx, gy = x, y }
		}
		//fmt.Printf("\n")
	}
	gxd := 0
	gyd := -1
	area[gy][gx] = 'X'

	for gx+gxd >= 0 && gx+gxd < xlen && gy+gyd >= 0 && gy+gyd < ylen {
		if area[gy+gyd][gx+gxd] == '#' {
			gxd, gyd = -gyd, gxd
		} else {
			gx += gxd
			gy += gyd
			area[gy][gx] = 'X'
		}
	}

	// fmt.Printf("\n")
	// for y := 0; y < ylen; y++ {
	// 	for x := 0; x < xlen; x++ {
	// 		fmt.Printf("%c", rune(area[y][x]))
	// 	}
	// 	fmt.Printf("\n")
	// }

	sum := 0
	for y := 0; y < ylen; y++ {
		for x := 0; x < xlen; x++ {
			if area[y][x] == 'X' { sum++ }
		}
	}
	fmt.Printf("= %v\n", sum)
	// 5161 too low
}

func star2() {
	inBytes,_ := os.ReadFile("06.txt")
	in := string(inBytes)

	area := make([][]byte, 0)
	
	for _,line := range strings.Split(in, "\n") {
		if strings.TrimSpace(line) != "" {
			area = append(area, []byte(strings.TrimSpace(line)))
		}
	}

	ylen := len(area)
	xlen := len(area[0])

	

	sx, sy := -1,-1
	for y := 0; y < ylen; y++ {
		for x := 0; x < xlen; x++ {
			//fmt.Printf("%c", rune(area[y][x]))
			if area[y][x] == '^' { sx, sy = x, y }
		}
		//fmt.Printf("\n")
	}

	loops := 0
	for oy := 0; oy < ylen; oy++ {
		for ox := 0; ox < xlen; ox++ {
			if area[oy][ox] != '.' { continue }
			area[oy][ox] = '#'
			gx, gy := sx, sy
			gxd := 0
			gyd := -1
			area[gy][gx] = 'X'
			steps := 0
			for steps < ylen*xlen && gx+gxd >= 0 && gx+gxd < xlen && gy+gyd >= 0 && gy+gyd < ylen {
				if area[gy+gyd][gx+gxd] == '#' {
					gxd, gyd = -gyd, gxd
				} else {
					gx += gxd
					gy += gyd
					area[gy][gx] = 'X'
				}
				steps++
			}
			if steps == ylen*xlen { loops++ }
			for y := 0; y < ylen; y++ {
				for x := 0; x < xlen; x++ {
					if area[y][x] != '#' { area[y][x] = '.' }
				}
			}
			area[oy][ox] = '.'
		}
	}
	

	// fmt.Printf("\n")
	// for y := 0; y < ylen; y++ {
	// 	for x := 0; x < xlen; x++ {
	// 		fmt.Printf("%c", rune(area[y][x]))
	// 	}
	// 	fmt.Printf("\n")
	// }

	fmt.Printf("= %v\n", loops)
}

func main() {
	star1()
	star2()
}
